Content based naming for skeleton types (#9457)
dotnet/csharplang: meetings/working-groups/extensions/content-based-naming.md

## Summary

> The current metadata strategy for skeleton assemblies relies on ordinals to group skeleton types generated from extension blocks. The use of ordinals causes friction in a few parts of the C# ecosystem: 

Reviewer: We never used the "skeleton" term for assemblies. In fact, we are not producing special assemblies for "Extension Everything" feature.

> The current metadata strategy for skeleton assemblies relies on ordinals to group skeleton types generated from extension blocks. The use of ordinals causes friction in a few parts of the C# ecosystem: 

Reviewer: I think the ordinal is used to name nested types representing extension blocks. We are not really grouping them.

- Edit and Continue: any use of ordinals tied to source ordering means that adding or removing elements can create unnecessary, and possibly unresolvable, conflicts during ENC operations. Natural operations like adding a new extension block in the middle of two others can lead to renaming of existing blocks which is difficult to reconcile.

> - Public API Tracking: the skeleton assemblies are necessarily generated as `public` APIs and there are many tools in the .NET ecosystem that track `public` API usage. The current design means moving types around in source can observably change the set of `public` API in an assembly. Shipping with this design would mean this ecosystem of tools would need to change to treat this `public` API as _special_ in some way.

Reviewer: Again a reference to a "skeleton assemblies", I assume you meant to refer to nested types representing extension blocks.

> - Public API Tracking: the skeleton assemblies are necessarily generated as `public` APIs and there are many tools in the .NET ecosystem that track `public` API usage. The current design means moving types around in source can observably change the set of `public` API in an assembly. Shipping with this design would mean this ecosystem of tools would need to change to treat this `public` API as _special_ in some way.

Reviewer: "moving types around" vs. "moving extension blocks around"

- CREF generate an observable reference to skeleton assemblies via `inheritdoc`. This means source reordering can break any place CREFs are treated as a durable item. 

This proposal wants to revamp our approach to skeleton assembly metadata such that the following goals are achieved:

> 1. The set of skeleton methods and types that will be generated for a given extension member can be determined by looking at solely that extension member. There is no need to consider other members in the same extension block, `partial` declarations in the containing type, etc ...

Reviewer: ```suggestion
1. The set of methods and types that will be generated for a given extension member can be determined by looking at solely that extension member. There is no need to consider other members in the same extension block, `partial` declarations in the containing type, etc ...
```

> 2. Skeleton assemblies are treated like `public`. This proposal accepts that skeleton types and members will be observable by design and build time processes. This means they need to have, as much as possible, the same characteristics as other `public` API. Specifically the following types of actions should **not** break ABI compatibility for `public` members in a skeleton type:

Reviewer: ```suggestion
2. Extension types are treated like `public`. This proposal accepts that extension types and members will be observable by design and build time processes. This means they need to have, as much as possible, the same characteristics as other `public` API. Specifically the following types of actions should **not** break ABI compatibility for `public` members in a skeleton type:
```

    1. Reordering members in source code

>     2. Changing C# specific aspects of the type like adding / removing nullable annotations, changing tuple names, etc ...

Reviewer: ```suggestion
    2. Changing C# specific aspects of the extended type like adding / removing nullable annotations, changing tuple names, etc ...
```

3. The C# compiler should be able to fully rehydrate the original extension block from metadata. This means for each skeleton member the compiler can recreate the _exact_ type name that was written in source. This includes items like the original type parameter names, nullable annotations, etc ...

> This will be achieved by moving to a content based naming scheme for skeleton types and members.

Reviewer: ```suggestion
This will be achieved by moving to a content based naming scheme for extension types and compiler generated members.
```

## Detailed Design

This proposal relies heavily on using content based naming for skeleton types and members.  The proposal is going to focus on what items are included in the content name, it will not discuss the actual name produced. That is because the specific hashing algorithm used will be an implementation detail of the compiler.

> For an extension block the content name of the skeleton type will be determined by using the following parts of the extension type declaration:

Reviewer: ```suggestion
For an extension block the content name of the extension type will be determined by using the following parts of the extension block declaration:
```

- The fully qualified CLR name of the type. This will include namespaces, type constraints, named constraints (like `new`). Constraints will be ordered in the content name such that reordering them in source code does not change the name. Specifically:
    - Type parameter constraints will be listed in declaration order. The constraints for the Nth type parameter will occur before the Nth+1 type parameter.
    - Base type and interface constraints will be sorted by comparing the full names ordinally
    - Non-type constraints will be sorted by comparing the C# text ordinally
- This will not include any C# isms like tuple names, nullability, etc ... 

> The type parameters of the skeleton type will match those of the extension block. The exception being that the name of the type parameters will be normalized to `T0`, `T1`, etc ... based on the order they appear in the type declaration. This means an extension block with a type parameter `Dictionary<TKey, TValue>` will result in a skeleton type with type parameters `T0` and `T1` in that order.

Reviewer: ```suggestion
The type parameters of the extension type will match those of the extension block. The exception being that the name of the type parameters will be normalized to `T0`, `T1`, etc ... based on the order they appear in the type declaration. This means an extension block with a type parameter `Dictionary<TKey, TValue>` will result in a skeleton type with type parameters `T0` and `T1` in that order.
```

This will achieve the goal that the name of skeleton types will only change when the underlying CLR type of an extension type changes. Any change to C# parts of the type such as adding nullable annotations or tuple names will impact the skeleton type name.

> For an extension block a `private` method, referred to as the source type method, will be generated in the skeleton type. This method will have a single parameter which mirrors the extension `this` parameter. Specifically it will have the same attributes, ref declaration, parameter name and C# type. The name of this method will be a content name that is determined by using the following parts of the extension parameter declaration:

Reviewer: ```suggestion
For an extension block a `private` method, referred to as the source type method, will be generated in the extension type. This method will have a single parameter which mirrors the extension `this` parameter. Specifically it will have the same attributes, ref declaration, parameter name and C# type. The name of this method will be a content name that is determined by using the following parts of the extension parameter declaration:
```

> - The fully qualified C# name of the type. This will include items like nullable annotations, tuple names, etc ... The constraints will have the same ordering as CLR types for the skeleton type name.

Reviewer: ```suggestion
- The fully qualified C# name of the extended type. This will include items like nullable annotations, tuple names, etc ... The constraints will have the same ordering as CLR types for the skeleton type name.
```

> - The fully qualified C# name of the type. This will include items like nullable annotations, tuple names, etc ... The constraints will have the same ordering as CLR types for the skeleton type name.

Reviewer: > The constraints will have the same ordering as CLR types for the skeleton type name.

This is probably better to have as a separate bullet point, explaining in more details since this method is not generic and it is not clear what constraints we are talking about. I assume you are referring to extension block type parameter constraint types, but now using C# type names.

    }

    extension<T>(IEnumerable<U> source)
        where T : IEquatable<U>
    {
        public bool IsPresent(U value) => source.Any(x => x.Equals(value));
    }
}
```

This would generate the following:

```cs
class E
{
    public class <>ContentName_For_IEnumerable_T<T0>
    {
        [TypeParameterNames("T")]
        private void <>ContentName1(IEnumerable<T0> source) { }

        [TypeParameterNamess("T")]
        private void <>ContentName2(ref IEnumerable<T0?> p) { }

        [SourceTypeMethod("ContentName1")]
        public bool IsEmpty => throw null!;

        [SourceTypeMethod("ContentName1")]
        public int Count => throw null!;
    
        [SourceTypeMethod("ContentName2")]
        public bool AnyNull => throw null!;
    
        [SourceTypeMethod("ContentName2")]
        public bool NullToEmpty() => throw null!;
    }
  
    public class <>ContentName_For_IEnumerable_T_With_Constraint<T0>
       where T : IEquatable<T0>
    {
        [TypeParameterNames("U")]
        private void <>ContentName3(IEnumerable<T0> source) { }

        [SourceTypeMethod("ContentName3")]
        public static bool IsPresent(T0 value) => throw null!;
    }
}
```

> This approach will result in stable names for our skeleton types and members that will make it much more consumable for the existing C# ecosystem.

Reviewer: ```suggestion
This approach will result in stable names for our extension types and compiler generated members that will make it much more consumable for the existing C# ecosystem.
```

- The name of the extension `this` parameter
- The ref-ness of the extension `this` parameter
- The fully qualified name + attribute arguments for any attributes applied to the extension `this` parameter.

> - The names of any type parametercs declared on the extension block. These will be sorted in declaration order.

Reviewer 2: ```suggestion
- The names of any type parameters declared on the extension block. These will be sorted in declaration order.
```
